[WTF] Add StructDump for debugging

Constellation opened the pull request:
#### 3c94ba6d70070d52a469c04ece16da7d0edb467b
<pre>
[WTF] Add StructDump for debugging
<a href="https://bugs.webkit.org/show_bug.cgi?id=291127">https://bugs.webkit.org/show_bug.cgi?id=291127</a>
<a href="https://rdar.apple.com/problem/148664792">rdar://problem/148664792</a>

Reviewed by Keith Miller.

This adds `structDump` debugging helper function for `dataLog`.
If you are using clang, then it dumps *any* kind of struct / class
fields without adding explicit dump code by using clang&apos;s __builtin_dump_struct.

For example, if you run the code like,

    DFG::Node* node = ...;
    dataLogLn(structDump(*node));

Then you can get the output like this.

    JSC::DFG::Node {
      NodeOrigin origin = *0x10631d800
      NodeType m_op = 59
      unsigned int m_index = 323
      AdjacencyList children = *0x10631d818
      VirtualRegister m_virtualRegister = *0x10631d830
      unsigned int m_refCount = 1
      NodeFlags m_flags = 344069
      SpeculatedType m_prediction = 206158430208
      OpInfoWrapper m_opInfo = *0x10631d848
      OpInfoWrapper m_opInfo2 = *0x10631d850
      union (unnamed) m_misc = {
        Node * replacement = 0x0
        unsigned int epoch = 0
      }
      BasicBlock * owner = 0x1061177b0
    }

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StructDump.h: Added.
(WTF::StructDump::StructDump):
(WTF::StructDump::dump const):
(WTF::structDump):

Canonical link: <a href="https://commits.webkit.org/293332@main">https://commits.webkit.org/293332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d11445c53f0cef3550ce83ad30a3f8b296921bbe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98598 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18231 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8467 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103719 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49169 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18524 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26682 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75053 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32205 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101602 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14050 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89046 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55411 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13830 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7011 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48568 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91285 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83785 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7089 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106092 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97227 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25688 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18713 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84027 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26065 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85247 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83512 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28156 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5837 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19378 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15979 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25646 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30828 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/120845 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25464 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33817 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28784 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27039 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WTF/wtf/StructDump.h:
```diff
@@ -0,0 +1,75 @@
+IGNORE_WARNINGS_END
+            });
+#else
+        out.print("<struct>");
```
Maybe we should do a `pointerDump()` for GCC?

Constellation replied:
That sounds nice! Changed.

kmiller68 commented on Source/WTF/wtf/StructDump.h:
```diff
@@ -0,0 +1,75 @@
+    void dump(PrintStream& out) const
+    {
+#if COMPILER(CLANG)
+        __builtin_dump_struct(&m_data,
```
Nice, this is awesome!

nullhook replied:
finally!

Constellation committed:
[WTF] Add StructDump for debugging
https://bugs.webkit.org/show_bug.cgi?id=291127
rdar://problem/148664792

Reviewed by Keith Miller.

This adds `structDump` debugging helper function for `dataLog`.
If you are using clang, then it dumps *any* kind of struct / class
fields without adding explicit dump code by using clang's __builtin_dump_struct.

For example, if you run the code like,

    DFG::Node* node = ...;
    dataLogLn(structDump(*node));

Then you can get the output like this.

    JSC::DFG::Node {
      NodeOrigin origin = *0x10631d800
      NodeType m_op = 59
      unsigned int m_index = 323
      AdjacencyList children = *0x10631d818
      VirtualRegister m_virtualRegister = *0x10631d830
      unsigned int m_refCount = 1
      NodeFlags m_flags = 344069
      SpeculatedType m_prediction = 206158430208
      OpInfoWrapper m_opInfo = *0x10631d848
      OpInfoWrapper m_opInfo2 = *0x10631d850
      union (unnamed) m_misc = {
        Node * replacement = 0x0
        unsigned int epoch = 0
      }
      BasicBlock * owner = 0x1061177b0
    }

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StructDump.h: Added.
(WTF::StructDump::StructDump):
(WTF::StructDump::dump const):
(WTF::structDump):

Canonical link: https://commits.webkit.org/293332@main
